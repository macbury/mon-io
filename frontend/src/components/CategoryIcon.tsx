import React, { useMemo } from 'react'
import LinearGradient from './LinearGradient'
import styled, { useTheme } from 'styled-components/native'
import Color from 'color'
import Icon from './Icon'

interface IContainerProps {
  size?: number
}

const GradientContainer = styled(LinearGradient)`
  margin-bottom: 10px;
  /* padding: 9px; */
  border-radius: ${(props : IContainerProps) => props.size / 2}px;
  width: ${(props : IContainerProps) => props.size}px;
  height: ${(props : IContainerProps) => props.size}px;
  border-width: 3px;
  border-color: ${(props : any) => props.color};
  overflow: hidden;
  display: flex;
  align-items: center;
  justify-content: center;
`

const SimpleContainer = styled.View`
  margin-bottom: 0px;
  margin-right: 15px;
  border-radius: 50px;
  width: ${(props : IContainerProps) => props.size};
  height: ${(props : IContainerProps) => props.size};
  background: ${(prop) => prop.theme.itemBorderColor};
  display: flex;
  align-items: center;
  justify-content: center;
`

export interface ICategoryIconProp extends IContainerProps {
  color: string;
  name: string;
  simple?: boolean;
}

function colorByTheme(color: string, power: number, dark: boolean) {
  let out = Color(color)
  if (dark) {
    return out.darken(power).toString()
  } else {
    return out.lighten(power).toString()
  }
}

function GradientCategoryIcon({ name, color, size, ...props } : ICategoryIconProp) {
  const { dark } = useTheme()
  const endColor = useMemo(() => colorByTheme(color, 0.3, dark), [color, dark])
  const borderColor = useMemo(() => colorByTheme(color, 0.4, dark), [color, dark])
  const colors = dark ? [color, endColor] : [endColor, color]

  if (!size) {
    size = 64;
  }

  return (
    <GradientContainer colors={colors} color={borderColor} size={size} {...props}>
      <Icon name={name} size={size/2} color="#fff" />
    </GradientContainer>
  )
}

function SimpleCategoryIcon({ name, size, ...props } : ICategoryIconProp) {
  return (
    <SimpleContainer size={size} {...props}>
      <Icon name={name} size={size/2} color="#fff" />
    </SimpleContainer>
  )
}

export default function CategoryIcon(props : ICategoryIconProp) {
  if (props.simple) {
    return <SimpleCategoryIcon {...props} />
  } else {
    return <GradientCategoryIcon {...props} />
  }
}