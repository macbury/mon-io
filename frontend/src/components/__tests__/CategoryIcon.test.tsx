import React from 'react'
import { shallow } from 'enzyme'
import snapshot from 'enzyme-to-json'
import CategoryIcon, { ICategoryIconProp } from '../CategoryIcon'

it('renders full version', () => {
  const props : ICategoryIconProp = {
    color: '#fff',
    name: 'repeat',
    simple: false,
    size: 34
  }

  const component = shallow(<CategoryIcon {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})

it('renders simple version', () => {
  const props : ICategoryIconProp = {
    color: '#ddd',
    name: 'users',
    simple: true,
    size: 24
  }

  const component = shallow(<CategoryIcon {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})