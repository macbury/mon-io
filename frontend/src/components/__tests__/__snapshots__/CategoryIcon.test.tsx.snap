// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders full version 1`] = `
<GradientCategoryIcon
  color="#fff"
  name="repeat"
  simple={false}
  size={34}
/>
`;

exports[`renders simple version 1`] = `
<SimpleCategoryIcon
  color="#ddd"
  name="users"
  simple={true}
  size={24}
/>
`;
